bluemira/equilibria/coils.py

```diff
@@ -85,634 +108,631 @@ def generate_name(coil: Any, idx: Optional[int] = None):
+        Notes
+        -----
+        Only discretisation method currently implemented is rectangular fraction
 
-        if name is None:
-            # We need to have a reasonable coil name
-            name = CoilNamer.generate_name(self, None)
-        self.name = name
+        """
+        weighting = None
+        self._quad_x = self._x.copy()
+        self._quad_dx = self._dx.copy()
+        self._quad_z = self._z.copy()
+        self._quad_dz = self._dz.copy()
 
-        self.__sizer = CoilSizer(self)
-        self.__sizer(self)
+        if d_coil is not None:
+            # How fancy do we want the mesh or just smaller rectangles?
+            weighting = self._rectangular_discretisation(d_coil)
 
-        self.sub_coils = None
+        if weighting is None:
+            self._einsum_str = "..., ...j -> ..."
+            self._quad_weighting = np.ones((self._x.shape[0], 1))
+        else:
+            self._einsum_str = "...j, ...j -> ..."
+            self._quad_weighting = weighting
 
-    @property
-    def n_coils(self):
+    def psi(self, x, z):
         """
-        Number of coils.
+        Calculate poloidal flux at (x, z)
         """
-        return 1
+        return self.control_psi(x, z) * self.current
 
-    def set_current(self, current):
+    def psi_greens(self, pgreen):
         """
-        Sets the current in a Coil object
-
-        Parameters
-        ----------
-        current: float
-            The current to set in the coil
+        Calculate plasma psi from Greens functions and current
         """
-        self.current = current
-        if self.sub_coils is not None:
-            for coil in self.sub_coils.values():
-                coil.set_current(current / len(self.sub_coils))
+        return self.current * pgreen
 
-    def adjust_current(self, d_current):
+    def control_psi(self, x, z):
         """
-        Adjusts the current in a Coil object
-
-        Parameters
-        ----------
-        d_current: float
-            The change in current to apply to the coil
+        Calculate poloidal flux at (x, z) due to a unit current
         """
-        self.current += d_current
-        if self.sub_coils is not None:
-            for coil in self.sub_coils.values():
-                coil.adjust_current(d_current / len(self.sub_coils))
+        x, z = np.ascontiguousarray(x), np.ascontiguousarray(z)
 
-    def set_position(self, x, z, dz=None):
+        return np.einsum(
+            self._einsum_str,
+            greens_psi(
+                self._quad_x[None],
+                self._quad_z[None],
+                x[..., None],
+                z[..., None],
+                self._quad_dx[None],
+                self._quad_dz[None],
+            ),
+            self._quad_weighting[None],
+        )
+
+    def Bx(self, x, z):
+        """
+        Calculate radial magnetic field Bx at (x, z)
         """
-        Sets the position of the Coil object
+        return self.control_Bx(x, z) * self.current
```

Comment:
I'm not sure here, but it looks like the Green's functions are weighted with a weight of 1 (?). So just from looking at this, it seems as though a discretised coil would have its full current multiplied at all the quadrature points? 

Update: It seems like this is not the case, though, so clearly I'm not following it right!

This einsum stuff I don't understand well, so I'm not sure if it's a dot product or a hadamard product... Is the padding to facilitate an einsum somehow?

Reply:
Einsum is all hadamard products with possible summations over specific axes. A dot product is just a hadamard product of two matrices where you sum over the 'middle' axes eg the last axis of the first array and the first axis of the last.

eg.

```python
In [14]: a = np.ones((4,3)) * 4 

In [15]: b = np.ones((3,4)) * 6

In [16]: np.dot(a, b)
Out[16]: 
array([[72., 72., 72., 72.],
       [72., 72., 72., 72.],
       [72., 72., 72., 72.],
       [72., 72., 72., 72.]])

In [17]: np.einsum("ij, jk -> ijk", a, b)
Out[17]: 
array([[[24., 24., 24., 24.],
        [24., 24., 24., 24.],
        [24., 24., 24., 24.]],

       [[24., 24., 24., 24.],
        [24., 24., 24., 24.],
        [24., 24., 24., 24.]],

       [[24., 24., 24., 24.],
        [24., 24., 24., 24.],
        [24., 24., 24., 24.]],

       [[24., 24., 24., 24.],
        [24., 24., 24., 24.],
        [24., 24., 24., 24.]]])

In [18]: np.einsum("ij, jk -> ik", a, b)
Out[18]: 
array([[72., 72., 72., 72.],
       [72., 72., 72., 72.],
       [72., 72., 72., 72.],
       [72., 72., 72., 72.]])
```

I'll comment on the padding in your other comment